Rust/src/lib.rs

```diff
@@ -1,111 +1,147 @@
-struct Parrot<'a> {
-    parrot_type: &'a str,
+const LOAD_FACTOR: f32 = 9.0;
+const BASE_SPEED: f32 = 12.0;
+const FIXED_BASE_SPEED: f32 = 24.0;
+
+pub struct ParrotConfig {
     number_of_coconuts: usize,
     voltage: f32,
     nailed: bool,
 }
 
-impl<'a> Parrot<'a> {
-    pub fn speed(&self) -> Result<f32, &'static str> {
-        match self.parrot_type {
-            "european_parrot" => Ok(base_speed()),
-            "african_parrot" => {
-                let african_speed = base_speed() - load_factor() * self.number_of_coconuts as f32;
-                if african_speed > 0.0 { Ok(african_speed) } else { Ok(0.0)}
-            }
-            "norwegian_blue_parrot" => {
-                if self.nailed == true {
-                    Ok(0.0)
-                }
-                else {
-                    Ok(compute_base_speed_for_voltage(self.voltage))
-                }
-            }
-            _ => Err("Should be unreachable!")
+impl ParrotConfig {
+    pub fn new(number_of_coconuts: usize, voltage: f32, nailed: bool) -> Self {
+        ParrotConfig {
+            number_of_coconuts: number_of_coconuts,
+            voltage: voltage,
+            nailed: nailed,
         }
     }
 }
 
-fn compute_base_speed_for_voltage(voltage: f32) -> f32 {
-    let fixed_base_speed = 24.0;
-    let base_speed_for_voltage = voltage * base_speed();
-    if base_speed_for_voltage < fixed_base_speed {
-        base_speed_for_voltage
+trait Parrot {
```

Comment:
Nice! I was thinking to do something like this also, but was put off by the repetition of having to implement the trait... should have stuck with it :+1:

Reply:
I did try to go full yolo for a while and use trait objects in the tests but fell into a world of pain im not ready for. :0)